File: src/main/java/olivermakesco/de/refmagic/registry/RefinedMagicBlocks.java
package olivermakesco.de.refmagic.registry;

import com.terraformersmc.terraform.sign.block.TerraformSignBlock;
import com.terraformersmc.terraform.sign.block.TerraformWallSignBlock;
import com.terraformersmc.terraform.wood.block.StrippableLogBlock;
import net.minecraft.block.*;
import net.minecraft.sound.BlockSoundGroup;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import olivermakesco.de.refmagic.Mod;
import olivermakesco.de.refmagic.block.*;
import org.quiltmc.qsl.block.extensions.api.QuiltBlockSettings;

public class RefinedMagicBlocks {
    public static Block kyriteOreEnd = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.END_STONE)
                    .sounds(BlockSoundGroup.STONE)
                    .requiresTool()
    );
    public static Block kyriteOreOverworld = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.STONE)
                    .sounds(BlockSoundGroup.STONE)
                    .requiresTool()
    );
    public static Block kyriteOreNether = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.NETHERRACK)
                    .sounds(BlockSoundGroup.NETHERRACK)
                    .requiresTool()
    );
    public static Block kyriteBlock = new Block(
            QuiltBlockSettings
                    .copyOf(Blocks.AMETHYST_BLOCK)
                    .requiresTool()
    );
    public static Block topazOre = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.GRANITE)
                    .sounds(BlockSoundGroup.STONE)
                    .requiresTool()
    );
    public static Block topazBlock = new Block(
            QuiltBlockSettings
                    .copyOf(Blocks.AMETHYST_BLOCK)
                    .requiresTool()
    );
    public static Block bismuthOreNether = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.BASALT)
                    .sounds(BlockSoundGroup.BASALT)
                    .requiresTool()
    );
    public static Block bismuthOreOverworld = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.DEEPSLATE)
                    .sounds(BlockSoundGroup.DEEPSLATE)
                    .requiresTool()
    );
    public static Block bismuthBlock = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.IRON_BLOCK)
                    .sounds(BlockSoundGroup.METAL)
                    .requiresTool()
    );
    public static Block citrineOre = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.BLACKSTONE)
                    .sounds(BlockSoundGroup.STONE)
                    .requiresTool()
    );
    public static Block citrineBlock = new Block(
            QuiltBlockSettings
                    .copyOf(Blocks.AMETHYST_BLOCK)
                    .sounds(BlockSoundGroup.AMETHYST_BLOCK)
                    .requiresTool()
    );
    public static Block opalOre = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.END_STONE)
                    .sounds(BlockSoundGroup.STONE)
                    .requiresTool()
    );
    public static Block opalBlock = new Block(
            QuiltBlockSettings
                    .copyOf(Blocks.AMETHYST_BLOCK)
                    .requiresTool()
    );
    public static Block roseQuartzOre = new RefinedMagicOreBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.END_STONE)
                    .sounds(BlockSoundGroup.STONE)
                    .requiresTool()
    );
    public static Block roseQuartzBlock = new Block(
            QuiltBlockSettings
                    .copyOf(Blocks.QUARTZ_BLOCK)
                    .sounds(BlockSoundGroup.AMETHYST_BLOCK)
                    .requiresTool()
    );
    public static Block altarBlock = new AltarTableBlock(
            QuiltBlockSettings
                    .copyOf(Blocks.SPRUCE_PLANKS)
    );

    public static Block strippedEnchantedStem = new PillarBlock(QuiltBlockSettings.copyOf(Blocks.CRIMSON_STEM));
    public static Block enchantedStem = new StrippableLogBlock(() -> strippedEnchantedStem, MapColor.PALE_PURPLE, QuiltBlockSettings.copyOf(Blocks.CRIMSON_STEM));
    public static Block strippedEnchantedHyphae = new PillarBlock(QuiltBlockSettings.copyOf(Blocks.CRIMSON_STEM));
    public static Block enchantedHyphae = new StrippableLogBlock(() -> strippedEnchantedHyphae, MapColor.PALE_PURPLE, QuiltBlockSettings.copyOf(Blocks.CRIMSON_STEM));
    public static Block enchantedPlanks = new Block(QuiltBlockSettings.copyOf(Blocks.WARPED_PLANKS));
    public static Block enchantedSlab = new SlabBlock(QuiltBlockSettings.copyOf(Blocks.WARPED_SLAB));
    public static Block enchantedStairs = new StairsBlock(enchantedPlanks.getDefaultState(), QuiltBlockSettings.copyOf(Blocks.WARPED_STAIRS));
    public static Block enchantedButton = new WoodenButtonBlock(QuiltBlockSettings.copyOf(Blocks.WARPED_BUTTON));
    public static Block enchantedPressurePlate = new PressurePlateBlock(PressurePlateBlock.ActivationRule.EVERYTHING, QuiltBlockSettings.copyOf(Blocks.WARPED_PRESSURE_PLATE));
    public static Block enchantedTrapdoor = new TrapdoorBlock(QuiltBlockSettings.copyOf(Blocks.WARPED_TRAPDOOR));
    public static Block enchantedDoor = new DoorBlock(QuiltBlockSettings.copyOf(Blocks.WARPED_DOOR));
    public static Block enchantedFence = new FenceBlock(QuiltBlockSettings.copyOf(Blocks.WARPED_FENCE));
    public static Block enchantedFenceGate = new FenceGateBlock(QuiltBlockSettings.copyOf(Blocks.WARPED_FENCE_GATE));
    public static TerraformSignBlock enchantedSign = new TerraformSignBlock(Mod.id("entity/sign/enchanted"), QuiltBlockSettings.copyOf(Blocks.WARPED_SIGN));
    public static Block enchantedWallSign = new TerraformWallSignBlock(Mod.id("entity/sign/enchanted"), QuiltBlockSettings.copyOf(Blocks.WARPED_WALL_SIGN).dropsLike(enchantedSign));

    public static Block enchantedWart = new Block(QuiltBlockSettings.copyOf(Blocks.NETHER_WART_BLOCK));

    public static Block enchantedShroomlight = new Block(QuiltBlockSettings.copyOf(Blocks.SHROOMLIGHT));

    public static Block enlium = new EnliumBlock(QuiltBlockSettings.copyOf(Blocks.END_STONE));

    public static Block enliumGrowth = new EnliumGrowthBlock();

    public static Block enchantedFungus = new MushroomPlantBlock(QuiltBlockSettings.copyOf(Blocks.CRIMSON_FUNGUS),() -> RefinedMagicWorldgen.enchantedFungusPlanted);

    public static Block hoopvine = new HoopvineBlock();

    public static void register() {
        register(kyriteOreEnd, Mod.id("kyrite_ore_end"));
        register(kyriteOreOverworld, Mod.id("kyrite_ore_overworld"));
        register(kyriteOreNether, Mod.id("kyrite_ore_nether"));
        register(kyriteBlock, Mod.id("kyrite_block"));
        register(topazOre, Mod.id("topaz_ore"));
        register(topazBlock, Mod.id("topaz_block"));
        register(bismuthOreNether, Mod.id("bismuth_ore_nether"));
        register(bismuthOreOverworld, Mod.id("bismuth_ore_overworld"));
        register(bismuthBlock, Mod.id("bismuth_block"));
        register(citrineOre, Mod.id("citrine_ore"));
        register(citrineBlock, Mod.id("citrine_block"));
        register(opalOre, Mod.id("opal_ore"));
        register(opalBlock, Mod.id("opal_block"));
        register(roseQuartzOre, Mod.id("rose_quartz_ore"));
        register(roseQuartzBlock, Mod.id("rose_quartz_block"));
        register(altarBlock, Mod.id("altar"));


        register(enlium, Mod.id("enlium"));
        register(enliumGrowth, Mod.id("enlium_growth"));
        register(hoopvine, Mod.id("hoopvine"));
        register(enchantedFungus, Mod.id("enchanted_fungus"));

        register(enchantedStem, Mod.id("enchanted_stem"));
        register(strippedEnchantedStem, Mod.id("stripped_enchanted_stem"));
        register(enchantedHyphae, Mod.id("enchanted_hyphae"));
        register(strippedEnchantedHyphae, Mod.id("stripped_enchanted_hyphae"));
        register(enchantedPlanks, Mod.id("enchanted_planks"));
        register(enchantedSlab, Mod.id("enchanted_slab"));
        register(enchantedStairs, Mod.id("enchanted_stairs"));
        register(enchantedButton, Mod.id("enchanted_button"));
        register(enchantedPressurePlate, Mod.id("enchanted_pressure_plate"));
        register(enchantedTrapdoor, Mod.id("enchanted_trapdoor"));
        register(enchantedDoor, Mod.id("enchanted_door"));
        register(enchantedFence, Mod.id("enchanted_fence"));
        register(enchantedFenceGate, Mod.id("enchanted_fence_gate"));
        register(enchantedSign, Mod.id("enchanted_sign"));
        register(enchantedWallSign, Mod.id("enchanted_wall_sign"));

        register(enchantedShroomlight, Mod.id("enchanted_shroomlight"));
        register(enchantedWart, Mod.id("enchanted_wart"));
    }

    public static void register(Block block, Identifier id) {
        Registry.register(Registry.BLOCK, id, block);
    }
}

File: src/main/kotlin/olivermakesco/de/refmagic/registry/RefinedMagicBlocks.kt
package olivermakesco.de.refmagic.registry

import com.terraformersmc.terraform.sign.block.TerraformSignBlock
import com.terraformersmc.terraform.sign.block.TerraformWallSignBlock
import com.terraformersmc.terraform.wood.block.StrippableLogBlock
import net.minecraft.block.*
import net.minecraft.sound.BlockSoundGroup
import net.minecraft.util.registry.Registry
import olivermakesco.de.refmagic.Mod
import olivermakesco.de.refmagic.block.*
import org.quiltmc.qsl.block.extensions.api.QuiltBlockSettings
import org.quiltmc.qkl.wrapper.minecraft.registry.*

object RefinedMagicBlocks {
    var kyriteOreEnd: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.END_STONE)
            .sounds(BlockSoundGroup.STONE)
            .requiresTool()
    )
    var kyriteOreOverworld: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.STONE)
            .sounds(BlockSoundGroup.STONE)
            .requiresTool()
    )
    var kyriteOreNether: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.NETHERRACK)
            .sounds(BlockSoundGroup.NETHERRACK)
            .requiresTool()
    )
    var kyriteBlock = Block(
        QuiltBlockSettings
            .copyOf(Blocks.AMETHYST_BLOCK)
            .requiresTool()
    )
    var topazOre: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.GRANITE)
            .sounds(BlockSoundGroup.STONE)
            .requiresTool()
    )
    var topazBlock = Block(
        QuiltBlockSettings
            .copyOf(Blocks.AMETHYST_BLOCK)
            .requiresTool()
    )
    var bismuthOreNether: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.BASALT)
            .sounds(BlockSoundGroup.BASALT)
            .requiresTool()
    )
    var bismuthOreOverworld: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.DEEPSLATE)
            .sounds(BlockSoundGroup.DEEPSLATE)
            .requiresTool()
    )
    var bismuthBlock: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.IRON_BLOCK)
            .sounds(BlockSoundGroup.METAL)
            .requiresTool()
    )
    var citrineOre: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.BLACKSTONE)
            .sounds(BlockSoundGroup.STONE)
            .requiresTool()
    )
    var citrineBlock = Block(
        QuiltBlockSettings
            .copyOf(Blocks.AMETHYST_BLOCK)
            .sounds(BlockSoundGroup.AMETHYST_BLOCK)
            .requiresTool()
    )
    var opalOre: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.END_STONE)
            .sounds(BlockSoundGroup.STONE)
            .requiresTool()
    )
    var opalBlock = Block(
        QuiltBlockSettings
            .copyOf(Blocks.AMETHYST_BLOCK)
            .requiresTool()
    )
    var roseQuartzOre: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.END_STONE)
            .sounds(BlockSoundGroup.STONE)
            .requiresTool()
    )
    var roseQuartzBlock = Block(
        QuiltBlockSettings
            .copyOf(Blocks.QUARTZ_BLOCK)
            .sounds(BlockSoundGroup.AMETHYST_BLOCK)
            .requiresTool()
    )
    var infestedNetherrack: Block = RefinedMagicOreBlock(
        QuiltBlockSettings
            .copyOf(Blocks.ANCIENT_DEBRIS)
            .sounds(BlockSoundGroup.NETHERRACK)
            .requiresTool()
    )
    var altarBlock: Block = AltarTableBlock(
        QuiltBlockSettings
            .copyOf(Blocks.SPRUCE_PLANKS)
    )
    var strippedEnchantedStem: Block = PillarBlock(
        QuiltBlockSettings
            .copyOf(Blocks.CRIMSON_STEM)
    )
    var enchantedStem: Block = StrippableLogBlock(
        { strippedEnchantedStem },
        MapColor.PALE_PURPLE,
        QuiltBlockSettings.copyOf(Blocks.CRIMSON_STEM)
    )
    var strippedEnchantedHyphae: Block = PillarBlock(
        QuiltBlockSettings
            .copyOf(Blocks.CRIMSON_STEM)
    )
    var enchantedHyphae: Block = StrippableLogBlock(
        { strippedEnchantedHyphae },
        MapColor.PALE_PURPLE,
        QuiltBlockSettings.copyOf(Blocks.CRIMSON_STEM)
    )
    var enchantedPlanks = Block(
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_PLANKS)
    )
    var enchantedSlab: Block = SlabBlock(
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_SLAB)
    )
    var enchantedStairs: Block = StairsBlock(
        enchantedPlanks.defaultState,
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_STAIRS)
    )
    var enchantedButton: Block = WoodenButtonBlock(
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_BUTTON)
    )
    var enchantedPressurePlate: Block = PressurePlateBlock(
        PressurePlateBlock.ActivationRule.EVERYTHING,
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_PRESSURE_PLATE)
    )
    var enchantedTrapdoor: Block = TrapdoorBlock(
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_TRAPDOOR)
    )
    var enchantedDoor: Block = DoorBlock(
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_DOOR)
    )
    var enchantedFence: Block = FenceBlock(
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_FENCE)
    )
    var enchantedFenceGate: Block = FenceGateBlock(
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_FENCE_GATE)
    )
    var enchantedSign = TerraformSignBlock(
        Mod.id("entity/sign/enchanted"),
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_SIGN)
    )
    var enchantedWallSign: Block = TerraformWallSignBlock(
        Mod.id("entity/sign/enchanted"),
        QuiltBlockSettings
            .copyOf(Blocks.WARPED_WALL_SIGN)
            .dropsLike(enchantedSign)
    )
    var enchantedWart = Block(
        QuiltBlockSettings
            .copyOf(Blocks.NETHER_WART_BLOCK)
    )
    var enchantedShroomlight = Block(
        QuiltBlockSettings
            .copyOf(Blocks.SHROOMLIGHT)
    )
    @JvmField
    var enlium: Block = EnliumBlock(
        QuiltBlockSettings
            .copyOf(Blocks.END_STONE)
    )
    var enliumGrowth: Block = EnliumGrowthBlock()
    var enchantedFungus: Block = MushroomPlantBlock(
        QuiltBlockSettings
            .copyOf(Blocks.CRIMSON_FUNGUS)
    ) {
        RefinedMagicWorldgen.enchantedFungusPlanted
    }
    var hoopvine: Block = HoopvineBlock()

    fun register() {
        Registry.BLOCK(Mod.modid) {
            kyriteOreEnd            withName "kyrite_ore_end"
            kyriteOreOverworld      withName "kyrite_ore_overworld"
            kyriteOreNether         withName "kyrite_ore_nether"
            kyriteBlock             withName "kyrite_block"
            topazOre                withName "topaz_ore"
            topazBlock              withName "topaz_block"
            bismuthOreNether        withName "bismuth_ore_nether"
            bismuthOreOverworld     withName "bismuth_ore_overworld"
            bismuthBlock            withName "bismuth_block"
            citrineOre              withName "citrine_ore"
            citrineBlock            withName "citrine_block"
            opalOre                 withName "opal_ore"
            opalBlock               withName "opal_block"
            roseQuartzOre           withName "rose_quartz_ore"
            roseQuartzBlock         withName "rose_quartz_block"
            infestedNetherrack      withName "infested_netherrack"
            altarBlock              withName "altar"
            enlium                  withName "enlium"
            enliumGrowth            withName "enlium_growth"
            hoopvine                withName "hoopvine"
            enchantedFungus         withName "enchanted_fungus"
            enchantedStem           withName "enchanted_stem"
            strippedEnchantedStem   withName "stripped_enchanted_stem"
            enchantedHyphae         withName "enchanted_hyphae"
            strippedEnchantedHyphae withName "stripped_enchanted_hyphae"
            enchantedPlanks         withName "enchanted_planks"
            enchantedSlab           withName "enchanted_slab"
            enchantedStairs         withName "enchanted_stairs"
            enchantedButton         withName "enchanted_button"
            enchantedPressurePlate  withName "enchanted_pressure_plate"
            enchantedTrapdoor       withName "enchanted_trapdoor"
            enchantedDoor           withName "enchanted_door"
            enchantedFence          withName "enchanted_fence"
            enchantedFenceGate      withName "enchanted_fence_gate"
            enchantedSign           withName "enchanted_sign"
            enchantedWallSign       withName "enchanted_wall_sign"
            enchantedShroomlight    withName "enchanted_shroomlight"
            enchantedWart           withName "enchanted_wart"
        }
    }
}